sql: update the cascader to use row containers

BramGruneir opened the pull request:
All datums in the cascader now use a row container. And this actually goes one step further and the queue itself no longer stores any datums and instead uses the same row containers that are used to store the rows needed for the final foreign key checking, further reducing the memory footprint.

This required pushing a bytesMonitor down into RowDeleter's DeleteRow. In this change, I've tried to minimize the number of changes to code that calls into the row deleter.

Release note: (sql) ON DELETE CASCADE operations are now memory bounded.

BramGruneir committed:
sql: update the cascader to use row containers

All datums in the cascader now use a row container. And this actually goes one step further and the queue itself no longer stores any datums and instead uses the same row containers that are used to store the rows needed for the final foreign key checking, further reducing the memory footprint.

This required pushing a bytesMonitor down into RowDeleter's DeleteRow. In this change, I've tried to minimize the number of changes to code that calls into the row deleter.

Release note: (sql) ON DELETE CASCADE operations are now memory bounded.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20706)
<!-- Reviewable:end -->

dt commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 9 of 9 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/sqlbase/cascader.go, line 343 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20706#-L0fhtVaPrs4CAXk3rVe:-L0fhtVaPrs4CAXk3rVf:b-r1zx5e) ([raw file](https://github.com/cockroachdb/cockroach/blob/3385a831883f95a07543937cd5f1abcd07a3b799/pkg/sql/sqlbase/cascader.go#L343)):*
> ```Go
> 
> 	// Delete all the rows in a new batch.
> 	deleteBatch := c.txn.NewBatch()
> ```

curious: what motivated moving this up?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20706#-:-L0fiVkIl7pde59T8Npl:bnfp4nl)*
<!-- Sent from Reviewable.io -->

knz commented:
General approach LGTM.
There's an error remaining (leak) and a few nits.

---

Reviewed 4 of 9 files at r1, 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/sql/delete.go, line 89 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20706#-L0ssDsP-gWFhFFm5Q0F:-L0ssDsQSDwc4oMom6nL:b-p60nt6) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ed772209f98e39b663c0973fe15015d0f3bb3fd/pkg/sql/delete.go#L89)):*
> ```Go
> 		alloc:      &p.alloc,
> 		mon:        &p.session.TxnState.mon,
> 	}
> ```

I'm :-1: on storing the monitor reference in the deleter/updater struct. Eventually we'll want to preserve the logical plans across transactions, and I already have started removing all references to transient states from the plans.

What you can do however, is pick up the monitor from the `params` struct passed during `startExec` / `Next`. 

---

*[pkg/sql/sqlbase/cascader.go, line 101 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20706#-L0stIdP3EBEjPIUeO78:-L0stIdP3EBEjPIUeO79:b-fhlaon) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ed772209f98e39b663c0973fe15015d0f3bb3fd/pkg/sql/sqlbase/cascader.go#L101)):*
> ```Go
> 			indexColIDs[referencingIndex.ColumnIDs[i]] = found
> 		} else {
> 			return roachpb.BatchRequest{}, errors.Errorf(
> ```

`pgerror.NewError(pgerror.Code....)`

---

*[pkg/sql/sqlbase/cascader.go, line 367 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20706#-L0sst0ezOc2nNx2sWz4:-L0sst0ezOc2nNx2sWz5:bd6yei1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ed772209f98e39b663c0973fe15015d0f3bb3fd/pkg/sql/sqlbase/cascader.go#L367)):*
> ```Go
> 		}
> 		if err := pkRowFetcher.StartScanFrom(ctx, &fetcher); err != nil {
> 			return nil, nil, 0, err
> ```

there's a leak here if you return without closing rowsChanged.

---

*[pkg/sql/sqlbase/cascader.go, line 372 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20706#-L0sswhANApFw6Leb0aY:-L0sswhANApFw6Leb0aZ:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ed772209f98e39b663c0973fe15015d0f3bb3fd/pkg/sql/sqlbase/cascader.go#L372)):*
> ```Go
> 			rowToDelete, _, _, err := pkRowFetcher.NextRowDecoded(ctx)
> 			if err != nil {
> 				return nil, nil, 0, err
> ```

ditto

---

*[pkg/sql/sqlbase/cascader.go, line 377 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20706#-L0ssxcnd5QcrS72GCQ6:-L0ssxcnd5QcrS72GCQ7:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ed772209f98e39b663c0973fe15015d0f3bb3fd/pkg/sql/sqlbase/cascader.go#L377)):*
> ```Go
> 			// Add the row to be checked for consistency changes.
> 			if _, err := rowsChanged.AddRow(ctx, rowToDelete); err != nil {
> 				return nil, nil, 0, err
> ```

ditto

---

*[pkg/sql/sqlbase/cascader.go, line 382 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20706#-L0ssyXFq8jcMmeGQ04w:-L0ssyXFq8jcMmeGQ04x:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ed772209f98e39b663c0973fe15015d0f3bb3fd/pkg/sql/sqlbase/cascader.go#L382)):*
> ```Go
> 			// Delete the row.
> 			if err := rowDeleter.deleteRowNoCascade(ctx, deleteBatch, rowToDelete, traceKV); err != nil {
> 				return nil, nil, 0, err
> ```

ditto

---

*[pkg/sql/sqlbase/cascader.go, line 389 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20706#-L0sszAzxPQegif_zSPQ:-L0sszAzxPQegif_zSPR:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ed772209f98e39b663c0973fe15015d0f3bb3fd/pkg/sql/sqlbase/cascader.go#L389)):*
> ```Go
> 	// Run the batch.
> 	if err := c.txn.Run(ctx, deleteBatch); err != nil {
> 		return nil, nil, 0, err
> ```

ditto

---

*[pkg/sql/sqlbase/cascader.go, line 525 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20706#-L0stRGZlNHvN2b5mUFL:-L0stRGZlNHvN2b5mUFM:b-t27zs8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ed772209f98e39b663c0973fe15015d0f3bb3fd/pkg/sql/sqlbase/cascader.go#L525)):*
> ```Go
> 		rowDeleter, exists := c.rowDeleters[tableID]
> 		if !exists {
> 			return errors.Errorf("programming error: could not find row deleter for table %d", tableID)
> ```

`pgerror.NewError(pgerror.CodeInternalError, ...)`

---

*[pkg/sql/sqlbase/rowwriter.go, line 929 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20706#-L0st9TlAiZqbcLCYR41:-L0st9TlAiZqbcLCYR42:b-t27zs8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ed772209f98e39b663c0973fe15015d0f3bb3fd/pkg/sql/sqlbase/rowwriter.go#L929)):*
> ```Go
> 	if rd.cascader != nil {
> 		if mon == nil {
> 			return errors.New("programming error: bytes monitor is nil")
> ```

`pgerror.NewError(pgerror.CodeInternalError, ...)`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20706#-:-L0stUSG62jJtBhIhv-7:b-hh2tdw)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
TFTR!  Made the requested changes.

---

Review status: 4 of 12 files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/sql/delete.go, line 89 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20706#-L0ssDsP-gWFhFFm5Q0F:-L0tvt8qXuy0EgISLNwE:bnn38hj) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ed772209f98e39b663c0973fe15015d0f3bb3fd/pkg/sql/delete.go#L89)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I'm :-1: on storing the monitor reference in the deleter/updater struct. Eventually we'll want to preserve the logical plans across transactions, and I already have started removing all references to transient states from the plans.

What you can do however, is pick up the monitor from the `params` struct passed during `startExec` / `Next`. 
</blockquote></details>

Done.  I've moved the monitor to be part of the init function for all tablewriters.  I think this is an improvement.  Now if only we had a monitor for schema changes...

---

*[pkg/sql/sqlbase/cascader.go, line 343 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20706#-L0fhtVaPrs4CAXk3rVe:-L0trTpYE9QjzI-sUvIb:b7uza1a) ([raw file](https://github.com/cockroachdb/cockroach/blob/3385a831883f95a07543937cd5f1abcd07a3b799/pkg/sql/sqlbase/cascader.go#L343)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

curious: what motivated moving this up?
</blockquote></details>

Hmm.  I needed to move it up a bit, that was too far.  Fixed.

---

*[pkg/sql/sqlbase/cascader.go, line 101 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20706#-L0stIdP3EBEjPIUeO78:-L0tqWaNFCoAgpGeXdQZ:bhxlxt) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ed772209f98e39b663c0973fe15015d0f3bb3fd/pkg/sql/sqlbase/cascader.go#L101)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`pgerror.NewError(pgerror.Code....)`
</blockquote></details>

I have no idea what the error code for this is, and it is the exacts same error that is in fk.go's line 623.  I'd be down for fixing both with the right guidance here.

---

*[pkg/sql/sqlbase/cascader.go, line 367 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20706#-L0sst0ezOc2nNx2sWz4:-L0trjpcMp7lQxLIEMof:b-m1fayf) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ed772209f98e39b663c0973fe15015d0f3bb3fd/pkg/sql/sqlbase/cascader.go#L367)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

there's a leak here if you return without closing rowsChanged.
</blockquote></details>

Nope.  rowsChanged is added to c.rowsChanged on the cascader that gets cleaned up when the cascadeAll returns (using a defer).

---

*[pkg/sql/sqlbase/cascader.go, line 372 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20706#-L0sswhANApFw6Leb0aY:-L0ts-W3hAkqXvaQdnCv:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ed772209f98e39b663c0973fe15015d0f3bb3fd/pkg/sql/sqlbase/cascader.go#L372)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

ditto

---

*[pkg/sql/sqlbase/cascader.go, line 377 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20706#-L0ssxcnd5QcrS72GCQ6:-L0ts0mazlHN2-_Bkd0X:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ed772209f98e39b663c0973fe15015d0f3bb3fd/pkg/sql/sqlbase/cascader.go#L377)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

ditto

---

*[pkg/sql/sqlbase/cascader.go, line 382 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20706#-L0ssyXFq8jcMmeGQ04w:-L0ts29Mbs9Qh0k9h6a1:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ed772209f98e39b663c0973fe15015d0f3bb3fd/pkg/sql/sqlbase/cascader.go#L382)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

ditto

---

*[pkg/sql/sqlbase/cascader.go, line 389 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20706#-L0sszAzxPQegif_zSPQ:-L0ts2hGs4xVe9Qcxfz0:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ed772209f98e39b663c0973fe15015d0f3bb3fd/pkg/sql/sqlbase/cascader.go#L389)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

ditto

---

*[pkg/sql/sqlbase/cascader.go, line 525 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20706#-L0stRGZlNHvN2b5mUFL:-L0ts4FE0jmuwv8eYYg-:bybe5ww) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ed772209f98e39b663c0973fe15015d0f3bb3fd/pkg/sql/sqlbase/cascader.go#L525)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`pgerror.NewError(pgerror.CodeInternalError, ...)`
</blockquote></details>

Cool.  Done. 

@jordanlewis suggested add ing "programming error", should I include that in the message?

---

*[pkg/sql/sqlbase/rowwriter.go, line 929 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20706#-L0st9TlAiZqbcLCYR41:-L0tsOV4x5OJc1mYfuUY:beii2ea) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ed772209f98e39b663c0973fe15015d0f3bb3fd/pkg/sql/sqlbase/rowwriter.go#L929)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`pgerror.NewError(pgerror.CodeInternalError, ...)`
</blockquote></details>

Done, and same question about "programming error".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20706#-:-L0u2-4oBRW3ElYLr8UQ:benc1fv)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Gentle ping to @knz for one final pass please!

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20706#-:-L10IVxqXSSTt8szeMPZ:balecs7)*
<!-- Sent from Reviewable.io -->

knz commented:
Yes this looks good now. See remaining comments.
You need to rebase and run the linter, we've changed the context.

---

Reviewed 8 of 8 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/sql/delete.go, line 89 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20706#-L0ssDsP-gWFhFFm5Q0F:-L1NLj_XLZ4jxlIus32-:b-7kf7ab) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ed772209f98e39b663c0973fe15015d0f3bb3fd/pkg/sql/delete.go#L89)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Done.  I've moved the monitor to be part of the init function for all tablewriters.  I think this is an improvement.  Now if only we had a monitor for schema changes...
</blockquote></details>

You can make a monitor for schema changes; in fact, I think it would be a good addition. Perhaps in a follow-up PR? At least, file an issue.

---

*[pkg/sql/sqlbase/cascader.go, line 101 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20706#-L0stIdP3EBEjPIUeO78:-L1NM7KGBO0r255ktGW2:b-mtgem8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ed772209f98e39b663c0973fe15015d0f3bb3fd/pkg/sql/sqlbase/cascader.go#L101)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

I have no idea what the error code for this is, and it is the exacts same error that is in fk.go's line 623.  I'd be down for fixing both with the right guidance here.
</blockquote></details>

CodeForeignKeyViolationError

Also possibly use `pgerror.NewErrorf` to avoid `Sprintf`

---

*[pkg/sql/sqlbase/cascader.go, line 367 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20706#-L0sst0ezOc2nNx2sWz4:-L1NMU7-CzpeUEsR2k2Q:b-q4eebe) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ed772209f98e39b663c0973fe15015d0f3bb3fd/pkg/sql/sqlbase/cascader.go#L367)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Nope.  rowsChanged is added to c.rowsChanged on the cascader that gets cleaned up when the cascadeAll returns (using a defer).
</blockquote></details>

ok thanks for explaining/reminding.

---

*[pkg/sql/sqlbase/cascader.go, line 525 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20706#-L0stRGZlNHvN2b5mUFL:-L1NM_6_zaeiejHiqX5E:bhzg7jp) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ed772209f98e39b663c0973fe15015d0f3bb3fd/pkg/sql/sqlbase/cascader.go#L525)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Cool.  Done. 

@jordanlewis suggested add ing "programming error", should I include that in the message?
</blockquote></details>

you can, it's not bad.

Also use `NewErrorf` to avoid the call to Sprintf.

---

*[pkg/sql/sqlbase/rowwriter.go, line 929 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20706#-L0st9TlAiZqbcLCYR41:-L1NMjlGx3XeEiFZxl-K:bb6mt9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ed772209f98e39b663c0973fe15015d0f3bb3fd/pkg/sql/sqlbase/rowwriter.go#L929)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Done, and same question about "programming error".
</blockquote></details>

:+1:

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20706#-:-L1NMm3pEO9_zb8jQAk9:b-j5tr8d)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/sql/sqlbase/cascader.go, line 101 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20706#-L0stIdP3EBEjPIUeO78:-L1Y8UuIBl5b-aMFP4Nz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ed772209f98e39b663c0973fe15015d0f3bb3fd/pkg/sql/sqlbase/cascader.go#L101)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

CodeForeignKeyViolationError

Also possibly use `pgerror.NewErrorf` to avoid `Sprintf`
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/cascader.go, line 525 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20706#-L0stRGZlNHvN2b5mUFL:-L1YkpUdfK05EFMD-PkZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ed772209f98e39b663c0973fe15015d0f3bb3fd/pkg/sql/sqlbase/cascader.go#L525)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

you can, it's not bad.

Also use `NewErrorf` to avoid the call to Sprintf.
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/rowwriter.go, line 929 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20706#-L0st9TlAiZqbcLCYR41:-L1Yl0DNOrqUVl-KIQQV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ed772209f98e39b663c0973fe15015d0f3bb3fd/pkg/sql/sqlbase/rowwriter.go#L929)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

:+1:
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20706)*
<!-- Sent from Reviewable.io -->
